model.ts

```diff
@@ -0,0 +1,266 @@
+		} catch (error) {
+			console.error(
+				`Error in searchByIdStateChange: ${
+					error instanceof Error ? error.message : error
+				}`
+			);
+		}
+	}
+	// Adjusts the available height based on window size and recalculates the number of rows.
+	adjustWindowSize(): void {
+		try {
+			if (typeof window === "undefined" || !window.innerHeight) {
+				throw new Error("Unable to access window dimensions");
+			}
+
+			// Determine the dynamic height of the header and pagination.
+			const mainHeadingElement = document.getElementById("main-heading");
+			const paginationElement = document.getElementById("pagination");
+
+			if (mainHeadingElement && paginationElement) {
+				this.headerHeight =
+					mainHeadingElement.clientHeight +
+					paginationElement.clientHeight;
+			} else {
+				console.error(
+					"Could not find main-heading and/or pagination elements"
+				);
+			}
+			if (!this.rowHeight) {
+				throw new Error("Row height is not properly configured");
+			}
+
+			this.availableHeight =
+				window.innerHeight - this.headerHeight - this.rowHeight * 2;
+			this.numRows = Math.floor(this.availableHeight / this.rowHeight);
+
+			if (this.numRows <= 0) {
+				console.log(
+					"Window size too small, setting minimum number of rows to 1"
+				);
+				this.numRows = 1;
+			}
+
+			// Calculating new values without modifying the state immediately.
+			let newFrom = this.from;
+			let newTo = this.from + this.numRows - 1;
+
+			// If it's the first set of records ("first page"), start from 0 and populate the whole window size.
+			if (this.from === 0) {
+				newFrom = 0;
+				newTo = this.numRows - 1;
+			}
+
+			// Ensure `newTo` doesn't exceed totalRecordCount and adjust `newFrom` accordingly,
+			// meaning populate the whole window size.
+			if (newTo >= this.totalRecordCount) {
+				newTo = this.totalRecordCount - 1;
+				newFrom = newTo - this.numRows + 1;
+			}
+
+			// Check if the highlighted ID is currently between from and to,
+			// too enable priority functionality (always visible in the window).
+			const highlightedId = this.getHighlightedId();
+			if (
+				highlightedId !== null &&
+				highlightedId >= this.from &&
+				highlightedId <= this.to
+			) {
+				// If newTo would be smaller than highlightedId, adjust to keep highlightedId in view.
+				if (newTo < highlightedId) {
+					newTo = highlightedId;
+					newFrom = newTo - this.numRows + 1;
+				}
+			}
+
+			// Now, after all conditions have been checked, set the state.
+			this.setFrom(newFrom);
+			this.setTo(newTo);
+		} catch (error) {
+			console.error(
+				`Error in adjustWindowSize: ${
+					error instanceof Error ? error.message : error
+				}`
+			);
+		}
+	}
+
+	async retrieveRecords(): Promise<void> {
+		try {
+			this.records = await this.apiManager.fetchRecords(
+				this.from,
+				this.to
+			);
+		} catch (error) {
+			console.error(
+				`Error retrieving records: ${
+					error instanceof Error ? error.message : error
+				}`
+			);
+		}
```

Comment:
Rather use `.catch` instead of the `try-catch` and make sure that you still throw an error, other wise calling functions don't know something went wrong.